quaternion_attitude_control/quaternion_attitude_control_test.py

```diff
@@ -0,0 +1,244 @@
+    return 2 * np.sign(qe[0]+1e-10) * np.array([qe[1], qe[2], qe[3]])
+
+def ftoq(f = np.array([0,0,1]), yaw = 0):
+    """
+    Calculate a desired attitude from 3D thrust and yaw in world frame
+    Assuming a default multicopter configuration.
+
+    Note: 0 yaw in world frame results in body x-axis projection
+    onto world x-y-plane being aligned with world x-axis.
+
+    Params:
+        f: [optional] desired 3D thrust vector in world frame. Defaults to upwards.
+        yaw: [optional] desired body yaw.
+
+    Returns:
+        attitude quaternion setpoint which meets the given goals
+    """
+    # align z-axis with thrust
+    body_z = f / la.norm(f)
+    # choose body x-axis orthogonal to z and with given yaw
+    yaw_direction = np.array([-sin(yaw), cos(yaw), 0])
+    body_x = np.cross(yaw_direction, body_z)
+    body_x /= la.norm(body_x)
+    # choose body y-axis right hand orthogonal to x- and z-axis
+    body_y = np.cross(body_z, body_x)
+    # assemble attitude matrix and quaternion from base vectors
+    Rd = np.column_stack((body_x, body_y, body_z))
+    return Quaternion(matrix=Rd)
+
+def dcm_z(q = Quaternion()):
+    """
+    Calculate the body z-axis base vector from body attitude.
+
+    Params:
+        q: [optional] body attitude quaternion. Defaults to level.
+
+    Returns:
+        unit base vector of body z-axis in world frame
+    """
+    # convert quaternion to rotation matrix
+    R = q.rotation_matrix
+    # take last column vector
+    return R[:,2]
+
+def vtoq(src = np.array([0,0,1]), dst = np.array([0,0,1]), eps = 1e-5):
+    """
+    Calculate quaternion representing the shortest rotation
+    from one vector to the other
+
+    Params:
+        src: [optional] Source 3D vector from which to start the rotation. Defaults to upwards direction.
+        dst: [optional] Destination 3D vector to which to rotate. Defaults to upwards direction.
+        eps: [optional] numerical thershold to catch 180 degree rotations. Defaults to 1e-5.
+
+    Returns:
+        quaternion rotationg the shortest way from src to dst
+    """
+    q = Quaternion()
+    cr = np.cross(src, dst)
+    dt = np.dot(src, dst)
+
+    if(la.norm(cr) < eps and dt < 0):
+        # handle corner cases with 180 degree rotations
+        # if the two vectors are parallel, cross product is zero
+        # if they point opposite, the dot product is negative
+        cr = np.abs(src)
+        if(cr[0] < cr[2]):
+            if(cr[0] < cr[2]):
+                cr = np.array([1,0,0])
+            else:
+                cr = np.array([0,0,1])
+        else:
+            if(cr[1] < cr[2]):
+                cr = np.array([0,1,0])
+            else:
+                cr = np.array([0,0,1])
+        q[0] = 0
+    else:
+        # normal case, do half-way quaternion solution
+        q[0] = dt + sqrt(np.dot(src,src) * np.dot(dst,dst))
+    q[1] = cr[0]
+    q[2] = cr[1]
+    q[3] = cr[2]
+    return q.normalised
+
+# setup 3D plot
+fig = plt.figure()

```

Comment:
Maybe we can add a title :smiley: 
`fig = plt.figure('quaternion attitude control')` or `plt.title('quaternion attitude control')`

Reply:
Makes sense, I build this in thanks.

Reply:
done